Restructure payouts file

fhenneke opened the pull request:
This PR is an attempt at implementing #427. The PR does not change the values of final results (up to rounding of floats). It changes the structure of the code and intermediate representations.

As a first step, it separates the computation of different parts of the accounting into different functions.

```python
    # compute individual components of payments
    solver_info = compute_solver_info(
        reward_target_df,
        service_fee_df,
        config,
    )
    rewards = compute_rewards(
        batch_data,
        quote_rewards_df,
        exchange_rate_native_to_cow,
        config.reward_config,
    )
    protocol_fees = compute_protocol_fees(batch_data)
    partner_fees = compute_partner_fees(batch_data, config.protocol_fee_config)
    buffer_accounting = compute_buffer_accounting(batch_data, slippage_df)
```
Those functions are implemented in separate files 

The results of these steps are converted into data frames for solver payments and for protocol and partner fee payments.

```python
    # combine into solver payouts and partner payouts
    solver_payouts = compute_solver_payouts(
        solver_info, rewards, protocol_fees, buffer_accounting
    )
    partner_payouts = (
        partner_fees  # no additional computation required here at the moment
    )
```

Payout data on transfers and overdrafts is then computed from solver and parner payouts data.

```python
    payouts = prepare_payouts(solver_payouts, partner_payouts, dune.period, config)
```

The code can be tested to produce the same transfer files as the old code.
Tests have been adapted and cover essentially what was tested before. There is a bit more strictness in the testing of the separate computations of rewards, protocol fees, partner fees, etc.
There is no end-to-end test for payments yet. This should be added at some point.

Future changes could remove data frames from intermediate results. This would make it easier to have correct types and detect and handle missing data. Data for the different parts of the accounting can be changed to use intermediate tables generated by `src/data_sync/sync_data.py`.

fhenneke committed:
added files for computing parts of the accounting

fhenneke committed:
use new functions in payouts

fhenneke committed:
update payouts test

fhenneke committed:
add tests for new files

fhenneke committed:
minor layout change

harisang committed:
Merge branch 'main' into restructure_payouts

harisang commented on src/fetch/payouts.py:
```diff
@@ -26,33 +32,21 @@
-PAYMENT_COLUMNS = {
+SOLVER_PAYOUTS_COLUMNS = [
     "solver",
+    "solver_name",
```
Do we need a name? Since we don't actually use names anywhere in our infra, maybe we can remove them altogether. Or does this have to do with the service_fee?

fhenneke replied:
We only use the name for overdrafts as far as I can tell. It should be removed at some point.

I did not remove it yet because I wanted to separate structural changes (this PR) from behavior changes.

harisang commented on src/fetch/buffer_accounting.py:
```diff
@@ -0,0 +1,52 @@
+) -> DataFrame:
+    """Compute buffer accounting per solver"""
+
+    # validate batch rewards and quote rewards columns
```
comment seems wrong

harisang commented on src/fetch/buffer_accounting.py:
```diff
@@ -0,0 +1,52 @@
+]
+
+
+def compute_buffer_accounting(
```
i guess not that much computation is done here. don't have a great alternative for a name, but i was expecting to actually compute the networks fees and do the slippage accounting within the function, and not just receive those as parameters

fhenneke replied:
This PR was supposed to be rather minimal and only changes the structure of the code (with minimal exceptions around default values).
It should then become easier to understand what data is required for different parts of the accounting and more of the actual computation can be moved into those functions.

harisang commented on src/fetch/partner_fees.py:
```diff
@@ -0,0 +1,62 @@
+PARTNER_FEES_COLUMNS = ["partner", "partner_fee_eth", "partner_fee_tax"]
+
+
+def compute_partner_fees(batch_data: DataFrame, config: ProtocolFeeConfig) -> DataFrame:
```
As there are many details about the accounting, i would suggest we add more explicit comments about what each function is computing and specifically, what its output is about.

e.g., here we could explicitly mention that:
- `partner_fee_eth` column is the full amount of partner fee in eth (or wei?), before CoW DAO claims a cut.
- `partner_fee_tax` is a percent and not an amount and corresponds to the percent of the `partner_fee_eth ` amount that CoW DAO claims.

fhenneke replied:
I added numpy style docstrings to all new files. This should cover the meaning of columns of inputs and outputs at least.

harisang commented on src/fetch/partner_fees.py:
```diff
@@ -0,0 +1,62 @@
+def compute_partner_fees(batch_data: DataFrame, config: ProtocolFeeConfig) -> DataFrame:
+    """Compute partner fees per integrator"""
+
+    # validate batch rewards and quote rewards columns
```
again seems like a copy-paste comment

harisang commented on src/fetch/protocol_fees.py:
```diff
@@ -0,0 +1,25 @@
+) -> DataFrame:
+    """Compute protocol fees per solver."""
+
+    # validate batch rewards and quote rewards columns
```
same here

harisang commented on src/fetch/rewards.py:
```diff
@@ -0,0 +1,77 @@
+
+log = set_log(__name__)
+
+BATCH_REWARDS_COLUMNS = ["solver", "primary_reward_eth"]
```
do you think we should drop the prefix `primary` ? Would be good to decide on some internal naming convention as well. Maybe (1) batch_auction_rewards and (2) quote rewards? I admit (1) is a bit long but since we are restructuring, would be good to remove any deprecated references to the primary/secondary rewards distinction we used to have

fhenneke replied:
I am typically referring to those different rewards as `batch_rewards` and `quote_rewards`.

I mostly did not rename it here since it should be renamed on the level of the sql query and that was beyond the scope of this PR. Once the respective database query is scpecific to computing rewards (as opposed to containing all kinds of stuff like partner fee addresses) it will be easier to rename a column in the query since the different parts of the accounting are handled in different functions.

harisang commented on src/fetch/rewards.py:
```diff
@@ -0,0 +1,77 @@
+
+def compute_rewards(
+    batch_data: DataFrame,
+    quote_rewards: DataFrame,
```
i would rename this to `quote_data`, to make it more uniform and also because there are no rewards stored in that dataframe

harisang replied:
Also, we could be more explicit by writing

- batch_data_per_solver
- quote_data_per_solver

fhenneke replied:
I am torn a bit between using more descriptive names (e.g. `batch_data_per_solver`) and more general names (e.g. `batch_data`).

On the one hand, in the end, the general code should not depend on what data exactly is needed to compute rewards, it just depends on `batch_data` and we choose what that entails. And the structure of that part of the code was originally intended to end up as something like
```python
rewards = compute_rewards(orderbook, dune, config)
```
where all the fetching and computing is abstracted away.

On the other hand, we will not end up with a clean version any time soon. So we might as well be as explicit as possible for now.

harisang commented on src/fetch/rewards.py:
```diff
@@ -0,0 +1,77 @@
+]
+
+
+def compute_rewards(
```
I was thinking whether we should express rewards in their "native currency", i.e., batch rewards in native token and quote rewards in COW, and then at the final step, when we build the payout, actually do the conversions.

This would mean dropping the `primary_reward_cow` column, as well as not pass the `exchange_rate` as a parameter.

fhenneke replied:
Originally, the intention was to have a data frame `rewards` containing all information for computing reward payments. This also works smoothly with the rest of the payments code. It does completely decouple the reward computation and the creation of payments.

I thought about relaxing this to being able to compute all reward information from  `rewards` and `reward_config`. This would allow dropping the strange `'reward_token_address'` column. But it would still require exchange rates to be part of that data frame.

harisang commented on src/fetch/solver_info.py:
```diff
@@ -0,0 +1,64 @@
+    service_fees: DataFrame,
+    config: AccountingConfig,
+) -> DataFrame:
+    """Compute solver information"""
```
A comment about the input, and specifically about the `service_fee` column (is it a boolean or a fraction), would help

harisang commented on src/fetch/solver_info.py:
```diff
@@ -0,0 +1,64 @@
+        for service_fees_flag in solver_info["service_fee"]
+    ]
+
+    if not solver_info["solver"].is_unique:
```
Can this happen as part of this function, or the only way this can happen is if the input table itself (`reward_targets`) contains duplicates?

fhenneke replied:
Good point. I'll will have another look at that. I do remember needing it at some point but those queries have changed a bit since then.
Potentially, this can be changed to an assert, and the actual changing of data happens in the dune fetching, as for other queries.
At some point, all the processing might take place in `compute_solver_info`, but for now we should keep it consistent.

harisang commented on src/fetch/payouts.py:
```diff
@@ -99,19 +93,15 @@ def __init__(  # pylint: disable=too-many-arguments, too-many-positional-argumen
         solver = frame["solver"]
         reward_target = frame["reward_target"]
-        if pandas.isna(reward_target):
+        if pd.isna(reward_target):
```
I would slightly edit the msg to write `Using solver submission address instead.`

harisang commented on src/fetch/payouts.py:
```diff
@@ -99,19 +93,15 @@ def __init__(  # pylint: disable=too-many-arguments, too-many-positional-argumen
-        if pandas.isna(buffer_accounting_target):
+        if pd.isna(buffer_accounting_target):
             log.warning(
                 f"Solver {solver} without buffer_accounting_target. Using solver"
```
Same comment as above

fhenneke committed:
address minor review comments

fhenneke committed:
rename column constants

fhenneke committed:
add documentation for new files

harisang commented on src/fetch/solver_info.py:
```diff
@@ -0,0 +1,119 @@
+            "0x"-prefixed hex representation of address of a solvers bonding pool.
+        solver_name: str
+            Name of a solver.
+    service_fees : DataFrame
```
Something is off with indentation here.

fhenneke replied:
I changed the format a bit to make it clearer what is an input and what is a column.

harisang commented on src/fetch/solver_info.py:
```diff
@@ -0,0 +1,119 @@
+        solver : str
+            "0x"-prefixed hex representation of the submission address of a solver.
+        service_fee : bool
+            True is a solver needs to pay a service fee to the COW DAO. Otherwise, it is False.
```
```suggestion
            True if a solver needs to pay a service fee to the COW DAO. Otherwise, it is False.
```

fhenneke committed:
fix bug with calling the vouch query twice

fhenneke committed:
fix typo in docstring

fhenneke committed:
add docstrings for payout functions

fhenneke committed:
change format of docstring

fhenneke commented:
I added some documentation for most functions which were added or changed.

I am trying to stick to a [numpy/scipy style docstring](https://numpydoc.readthedocs.io/en/latest/format.html) style. Additionally, I added explanations for all expected columns.

harisang commented on src/fetch/payouts.py:
```diff
@@ -339,160 +328,374 @@ def prepare_transfers(  # pylint: disable=too-many-arguments, too-many-positiona
+            Results of the buffer accounting are sent to this address. It is equal to `solver` or
+            `reward_target`.
+        - service_fee : Fraction
+            The fraction of rewards which need to be paid to the COW DAO.
```
```suggestion
            The fraction of rewards which need to be paid to the CoW DAO.
```

harisang commented on src/fetch/payouts.py:
```diff
@@ -339,160 +328,374 @@ def prepare_transfers(  # pylint: disable=too-many-arguments, too-many-positiona
+        solver_payouts["reward_token_address"] = (
+            solver_payouts["reward_token_address"]
+            .fillna(
+                "0x0000000000000000000000000000000000000001"
```
Is this ever needed btw?

harisang replied:
We could alternatively add an `assert` here as otherwise it the dataframe doesn't make sense (unless the corresponding row contains all zeros).

fhenneke replied:
This is quite a bit of code smell. I added it because without this default there were errors. But I do not exactly know why.

The reward token address is a required field in the `RewardAndPenaltyDatum` class and will be turned into an `Address`. The address is, however, never used for solvers who do not get a reward. In principle, whenever a solver settled an auction or provided a quote, they should have an entry in `rewards` and thus filling the in defaults should not be required.

I will have to have another look (or follow up on this in another PR).

fhenneke replied:
I removed this default value and adapted one of the tests accordingly.

harisang reviewed (approved):
Looks good! There is one small comment to be addressed but not a deal breaker

fhenneke committed:
correct spelling of CoW DAO

fhenneke committed:
reverse change in meaning of payout summary

fhenneke committed:
remove dummy cow token address

also adapt corresponding tests
